compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1507,23 +1510,26 @@ public void visit(BLangWhile whileNode, AnalyzerData data) {
         data.failureHandled = failureHandled;
         data.loopWithinTransactionCheckStack.pop();
         analyzeExpr(whileNode.expr, data);
-        analyzeOnFailClause(whileNode.onFailClause, data);
-        data.errorTypes.pop();
+        if (onFailExists) {
+            whileNode.body.failureBreakMode = getPossibleBreakMode(!data.errorTypes.peek().isEmpty());
+            analyzeOnFailClause(whileNode.onFailClause, data);
```

Comment:
Better to extract this into a method since they are multiple use cases

Reply:
Updated in daeb71d752dd2b0ec1b555a64009727b164a51b2